File: backend/processors/csv_processor.py
"""
CSV Processor Module

Registers S3-hosted CSV/Parquet files as metadata entries for the in-memory
DuckDB connector.  No local files are created — DuckDB is used only as a
temporary engine to read column/row metadata from S3 at registration time.

S3 storage layout (enforced by frontend upload-url route):
    {companyId}/csvs/{mode}/{connectionName}/{uuid}.csv   (or .parquet)

The company_id prefix ensures strict cross-company isolation.
"""

import io
import re
import uuid
import asyncio
from pathlib import Path
from typing import List, Dict, Any

import boto3
import duckdb
import pandas as pd
from config import (
    BASE_DUCKDB_DATA_PATH,
    OBJECT_STORE_BUCKET,
    OBJECT_STORE_REGION,
    OBJECT_STORE_ACCESS_KEY_ID,
    OBJECT_STORE_SECRET_ACCESS_KEY,
    OBJECT_STORE_ENDPOINT,
)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def sanitize_table_name(filename: str) -> str:
    """Convert a filename to a valid SQL table name.

    - Remove file extension
    - Replace spaces/hyphens with underscores
    - Strip characters that aren't alphanumeric or underscore
    - Lowercase
    - Prefix with 't_' if it starts with a digit
    - Truncate to 63 chars (PostgreSQL identifier limit)
    """
    name = Path(filename).stem
    name = name.replace(' ', '_').replace('-', '_')
    name = re.sub(r'[^a-zA-Z0-9_]', '', name)
    name = name.lower()
    if name and name[0].isdigit():
        name = 't_' + name
    if not name:
        name = 'table_data'
    return name[:63]


def ensure_unique_table_names(filenames: List[str]) -> Dict[str, str]:
    """Return {filename: table_name} with collision-free names within the list."""
    result: Dict[str, str] = {}
    used: set = set()
    for filename in filenames:
        base = sanitize_table_name(filename)
        final = base
        counter = 1
        while final in used:
            final = f"{base}_{counter}"[:63]
            counter += 1
        result[filename] = final
        used.add(final)
    return result


def detect_file_format(filename: str) -> str:
    ext = filename.rsplit('.', 1)[-1].lower() if '.' in filename else ''
    if ext in ('parquet', 'pq'):
        return 'parquet'
    if ext == 'xlsx':
        return 'xlsx'
    return 'csv'


def _open_s3_duckdb() -> duckdb.DuckDBPyConnection:
    """Open a temporary in-memory DuckDB connection configured for S3 access."""
    conn = duckdb.connect()
    conn.execute("INSTALL httpfs; LOAD httpfs;")
    conn.execute(f"SET s3_region='{OBJECT_STORE_REGION}'")
    conn.execute(f"SET s3_access_key_id='{OBJECT_STORE_ACCESS_KEY_ID or ''}'")
    conn.execute(f"SET s3_secret_access_key='{OBJECT_STORE_SECRET_ACCESS_KEY or ''}'")
    if OBJECT_STORE_ENDPOINT:
        conn.execute(f"SET s3_endpoint='{OBJECT_STORE_ENDPOINT}'")
        conn.execute("SET s3_url_style='path'")
    return conn


def _make_boto3_kwargs() -> Dict[str, Any]:
    kwargs: Dict[str, Any] = dict(
        region_name=OBJECT_STORE_REGION,
        aws_access_key_id=OBJECT_STORE_ACCESS_KEY_ID or None,
        aws_secret_access_key=OBJECT_STORE_SECRET_ACCESS_KEY or None,
    )
    if OBJECT_STORE_ENDPOINT:
        kwargs['endpoint_url'] = OBJECT_STORE_ENDPOINT
    return kwargs


def _expand_xlsx_to_csvs(
    s3_key: str,
    connection_name: str,
    company_id: int,
    mode: str,
    schema_name: str,
) -> List[Dict[str, str]]:
    """
    Download an xlsx from S3, convert each non-empty sheet to a CSV, upload the
    CSVs back to S3 under the same connection prefix, and return a list of file
    records ({filename, s3_key, schema_name, file_format}) — one per sheet.

    The original xlsx object is left in S3 (it will be removed with the
    connection when the user deletes it via the normal cleanup flow).
    """
    s3 = boto3.client('s3', **_make_boto3_kwargs())

    # Download xlsx bytes
    response = s3.get_object(Bucket=OBJECT_STORE_BUCKET, Key=s3_key)
    xlsx_bytes = response['Body'].read()

    # Parse with pandas
    xl = pd.ExcelFile(io.BytesIO(xlsx_bytes), engine='openpyxl')
    csv_records: List[Dict[str, str]] = []

    for sheet_name in xl.sheet_names:
        df = pd.read_excel(xl, sheet_name=sheet_name)
        if df.empty:
            continue
        safe_name = re.sub(r'[^a-zA-Z0-9_-]', '_', sheet_name)[:50].lower()
        csv_filename = f"{safe_name}.csv"
        csv_key = f"{company_id}/csvs/{mode}/{connection_name}/{uuid.uuid4().hex}_{csv_filename}"

        buf = io.StringIO()
        df.to_csv(buf, index=False)
        csv_bytes = buf.getvalue().encode('utf-8')

        s3.put_object(
            Bucket=OBJECT_STORE_BUCKET,
            Key=csv_key,
            Body=csv_bytes,
            ContentType='text/csv',
        )

        csv_records.append({
            'filename': csv_filename,
            's3_key': csv_key,
            'schema_name': schema_name,
            'file_format': 'csv',
        })

    if not csv_records:
        raise ValueError("No non-empty sheets found in the uploaded xlsx file")

    return csv_records


def _read_file_metadata(
    conn: duckdb.DuckDBPyConnection,
    s3_url: str,
    file_format: str,
    view_name: str,
) -> tuple:
    """Create a temp view and return (row_count, columns)."""
    if file_format == 'parquet':
        reader = f"read_parquet('{s3_url}')"
    else:
        reader = f"read_csv_auto('{s3_url}')"

    conn.execute(f'CREATE OR REPLACE TEMP VIEW "{view_name}" AS SELECT * FROM {reader}')

    row_count = conn.execute(f'SELECT COUNT(*) FROM "{view_name}"').fetchone()[0]

    # DESCRIBE returns: column_name, column_type, null, key, default, extra
    cols = conn.execute(f'DESCRIBE "{view_name}"').fetchall()
    columns = [{"name": c[0], "type": c[1]} for c in cols]
    return row_count, columns


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

async def process_csv_from_s3(
    company_id: int,
    mode: str,
    connection_name: str,
    files: List[Dict[str, str]],
    replace_existing: bool = False,
) -> Dict[str, Any]:
    """
    Validate S3-hosted CSV/Parquet files and collect metadata.

    No local files or DuckDB database files are created.  A temporary
    in-memory DuckDB is used only to read column names/types and row counts.

    Company isolation: S3 keys are expected to be prefixed with company_id
    (enforced by the frontend upload-url route).

    Args:
        company_id: Used only for logging/validation; the S3 key already
                    encodes the company so cross-company access is impossible.
        mode:       'org' | 'tutorial' etc.
        connection_name: Name of the CSV connection document.
        files:      List of {filename, s3_key, schema_name?, file_format?}
        replace_existing: Ignored (no-op); kept for API compatibility.

    Returns:
        {"files": [{table_name, schema_name, s3_key, file_format,
                    filename, row_count, columns}]}
    """
    if not OBJECT_STORE_BUCKET:
        raise ValueError(
            "OBJECT_STORE_BUCKET is not configured. "
            "Set OBJECT_STORE_BUCKET env var to enable S3-backed CSV connections."
        )

    # Expand xlsx files into per-sheet CSV records (blocking I/O runs off the event loop)
    expanded_files: List[Dict[str, str]] = []
    for file_info in files:
        fmt = file_info.get('file_format') or detect_file_format(file_info['filename'])
        if fmt == 'xlsx':
            schema_name = file_info.get('schema_name') or 'public'
            sheet_records = await asyncio.get_event_loop().run_in_executor(
                None,
                _expand_xlsx_to_csvs,
                file_info['s3_key'],
                connection_name,
                company_id,
                mode,
                schema_name,
            )
            expanded_files.extend(sheet_records)
        else:
            expanded_files.append(file_info)
    files = expanded_files

    # Separate files with user-provided table names from those that need auto-generation
    auto_gen_schema_filenames: Dict[str, List[str]] = {}
    for f in files:
        if not f.get('table_name'):
            schema = f.get('schema_name', 'public') or 'public'
            auto_gen_schema_filenames.setdefault(schema, []).append(f['filename'])

    schema_table_names: Dict[str, Dict[str, str]] = {
        schema: ensure_unique_table_names(fnames)
        for schema, fnames in auto_gen_schema_filenames.items()
    }

    # Validate user-provided names: unique within each schema, no collision with auto-gen names
    user_provided: Dict[str, set] = {}
    for f in files:
        if f.get('table_name'):
            schema = f.get('schema_name', 'public') or 'public'
            t = sanitize_table_name(f['table_name'])
            user_provided.setdefault(schema, set())
            if t in user_provided[schema]:
                raise ValueError(
                    f"Duplicate table name '{t}' in schema '{schema}'. "
                    "Please use unique table names."
                )
            user_provided[schema].add(t)

    # Cross-check: user-provided names must not collide with auto-generated names
    for schema, auto_names_map in schema_table_names.items():
        auto_names = set(auto_names_map.values())
        for t in user_provided.get(schema, set()):
            if t in auto_names:
                raise ValueError(
                    f"Table name '{t}' in schema '{schema}' conflicts with an "
                    "auto-generated name. Please rename the file or provide a unique table name."
                )

    conn = _open_s3_duckdb()
    file_metadata = []
    try:
        for file_info in files:
            filename = file_info['filename']
            s3_key = file_info['s3_key']
            schema_name = file_info.get('schema_name') or 'public'
            file_format = file_info.get('file_format') or detect_file_format(filename)
            # Use user-provided table name if present, else auto-generate
            if file_info.get('table_name'):
                table_name = sanitize_table_name(file_info['table_name'])
            else:
                table_name = schema_table_names[schema_name][filename]
            s3_url = f"s3://{OBJECT_STORE_BUCKET}/{s3_key}"

            # Unique view name to avoid conflicts between iterations
            view_name = f"_tmp_{schema_name}_{table_name}"
            row_count, columns = _read_file_metadata(conn, s3_url, file_format, view_name)

            file_metadata.append({
                "filename": filename,
                "table_name": table_name,
                "schema_name": schema_name,
                "s3_key": s3_key,
                "file_format": file_format,
                "row_count": row_count,
                "columns": columns,
            })
    finally:
        conn.close()

    return {"files": file_metadata}


def delete_csv_connection(company_id: int, mode: str, connection_name: str) -> bool:
    """
    Delete a CSV connection's S3 data.

    Lists and deletes all objects under the connection's S3 prefix.
    Returns True if any objects were deleted, False if none were found.
    """
    if not OBJECT_STORE_BUCKET:
        return False

    s3 = boto3.client('s3', **_make_boto3_kwargs())
    prefix = f"{company_id}/csvs/{mode}/{connection_name}/"

    paginator = s3.get_paginator('list_objects_v2')
    keys_to_delete = []
    for page in paginator.paginate(Bucket=OBJECT_STORE_BUCKET, Prefix=prefix):
        for obj in page.get('Contents', []):
            keys_to_delete.append({'Key': obj['Key']})

    if not keys_to_delete:
        return False

    s3.delete_objects(
        Bucket=OBJECT_STORE_BUCKET,
        Delete={'Objects': keys_to_delete},
    )
    return True


async def seed_company_tutorial(
    company_id: int,
    mode: str,
    connection_name: str,
    schema_name: str = "main",
) -> Dict[str, Any]:
    """
    Seed a company's tutorial CSV connection from the local mxfood.duckdb source.

    Reads each table from mxfood.duckdb and writes a Parquet file directly to S3
    using DuckDB's httpfs extension (no intermediate local files).  Returns the
    same file-metadata structure as process_csv_from_s3 so the caller can update
    the connection document directly.

    S3 key layout:
        {company_id}/csvs/{mode}/{connection_name}/{table_name}.parquet

    No-ops gracefully when:
      - OBJECT_STORE_BUCKET is not configured
      - mxfood.duckdb cannot be found
    """
    if not OBJECT_STORE_BUCKET:
        return {"files": [], "skipped": True, "reason": "OBJECT_STORE_BUCKET not configured"}

    mxfood_path = Path(BASE_DUCKDB_DATA_PATH) / "data" / "mxfood.duckdb"
    if not mxfood_path.exists():
        return {"files": [], "skipped": True, "reason": f"mxfood.duckdb not found at {mxfood_path}"}

    def _seed_sync() -> Dict[str, Any]:
        # Open source DuckDB and configure httpfs for S3 writes in one connection.
        conn = duckdb.connect(str(mxfood_path), read_only=True)
        try:
            conn.execute("INSTALL httpfs; LOAD httpfs;")
            conn.execute(f"SET s3_region='{OBJECT_STORE_REGION}'")
            conn.execute(f"SET s3_access_key_id='{OBJECT_STORE_ACCESS_KEY_ID or ''}'")
            conn.execute(f"SET s3_secret_access_key='{OBJECT_STORE_SECRET_ACCESS_KEY or ''}'")
            if OBJECT_STORE_ENDPOINT:
                conn.execute(f"SET s3_endpoint='{OBJECT_STORE_ENDPOINT}'")
                conn.execute("SET s3_url_style='path'")

            tables = [row[0] for row in conn.execute("SHOW TABLES").fetchall()]
            file_metadata = []

            for table_name in tables:
                s3_key = f"{company_id}/csvs/{mode}/{connection_name}/{table_name}.parquet"
                s3_url = f"s3://{OBJECT_STORE_BUCKET}/{s3_key}"

                # Write table to S3 as Parquet directly from the source DuckDB
                conn.execute(
                    f"COPY (SELECT * FROM main.\"{table_name}\") "
                    f"TO '{s3_url}' (FORMAT PARQUET)"
                )

                # Read metadata back from the newly written Parquet file
                view_name = f"_seed_{table_name}"
                row_count, columns = _read_file_metadata(conn, s3_url, "parquet", view_name)

                file_metadata.append({
                    "filename": f"{table_name}.parquet",
                    "table_name": table_name,
                    "schema_name": schema_name,
                    "s3_key": s3_key,
                    "file_format": "parquet",
                    "row_count": row_count,
                    "columns": columns,
                })
        finally:
            conn.close()

        return {"files": file_metadata}

    # Run the blocking DuckDB work off the event loop
    return await asyncio.get_event_loop().run_in_executor(None, _seed_sync)

File: frontend/lib/csv-processor.ts
import 'server-only';

/**
 * CSV and Google Sheets processing — all logic lives here in Node.js.
 * Handles xlsx expansion, S3 uploads/deletes, and DuckDB metadata extraction.
 */

import { randomUUID } from 'crypto';
import { writeFileSync, readFileSync, unlinkSync, mkdirSync, rmSync } from 'fs';
import { tmpdir } from 'os';
import { dirname, join, resolve } from 'path';
import {
  S3Client,
  GetObjectCommand,
  ListObjectsV2Command,
  DeleteObjectsCommand,
  DeleteObjectCommand,
} from '@aws-sdk/client-s3';
import { DuckDBInstance } from '@duckdb/node-api';
import * as XLSX from 'xlsx';
import {
  OBJECT_STORE_BUCKET,
  OBJECT_STORE_REGION,
  OBJECT_STORE_ACCESS_KEY_ID,
  OBJECT_STORE_SECRET_ACCESS_KEY,
  OBJECT_STORE_ENDPOINT,
  LOCAL_UPLOAD_PATH,
} from '@/lib/config';
import { createObjectStore, isLocalObjectStore } from '@/lib/object-store';
import { sanitizeTableName, ensureUniqueTableNames } from '@/lib/csv-utils';

// Re-export shared utilities so existing callers don't break
export { sanitizeTableName, ensureUniqueTableNames } from '@/lib/csv-utils';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface IncomingFile {
  filename: string;
  s3_key: string;
  schema_name?: string;
  file_format?: 'csv' | 'parquet' | 'xlsx';
  table_name?: string;
}

export interface RegisteredFile {
  table_name: string;
  schema_name: string;
  s3_key: string;
  file_format: 'csv' | 'parquet';
  filename: string;
  row_count: number;
  columns: Array<{ name: string; type: string }>;
}

// ─── Storage helpers ──────────────────────────────────────────────────────────

/** Returns the DuckDB-readable URL/path for a stored key. */
function getStorageUrl(key: string): string {
  if (!isLocalObjectStore()) return `s3://${OBJECT_STORE_BUCKET!}/${key}`;
  return resolve(join(LOCAL_UPLOAD_PATH, key));
}

// ─── S3 client ───────────────────────────────────────────────────────────────

function makeS3(): S3Client {
  if (!OBJECT_STORE_BUCKET) throw new Error('OBJECT_STORE_BUCKET is not configured');
  return new S3Client({
    region: OBJECT_STORE_REGION,
    credentials:
      OBJECT_STORE_ACCESS_KEY_ID && OBJECT_STORE_SECRET_ACCESS_KEY
        ? { accessKeyId: OBJECT_STORE_ACCESS_KEY_ID, secretAccessKey: OBJECT_STORE_SECRET_ACCESS_KEY }
        : undefined,
    ...(OBJECT_STORE_ENDPOINT ? { endpoint: OBJECT_STORE_ENDPOINT, forcePathStyle: true } : {}),
  });
}

// ─── Table name helpers ───────────────────────────────────────────────────────
// (implementations live in lib/csv-utils.ts — re-exported above for back-compat)

// ─── xlsx expansion ───────────────────────────────────────────────────────────

/**
 * Download an xlsx from S3, expand each non-empty sheet to a CSV, upload back to S3.
 * Returns one IncomingFile record per sheet.
 */
async function getStoredFileBytes(key: string): Promise<Buffer> {
  if (isLocalObjectStore()) {
    return readFileSync(resolve(join(LOCAL_UPLOAD_PATH, key)));
  }
  const s3 = makeS3();
  const response = await s3.send(new GetObjectCommand({ Bucket: OBJECT_STORE_BUCKET!, Key: key }));
  const chunks: Uint8Array[] = [];
  for await (const chunk of response.Body as AsyncIterable<Uint8Array>) chunks.push(chunk);
  return Buffer.concat(chunks);
}

async function expandXlsxFromS3(
  s3Key: string,
  connectionName: string,
  companyId: number,
  mode: string,
  schemaName: string,
  createdKeys: string[],
): Promise<IncomingFile[]> {
  const buffer = await getStoredFileBytes(s3Key);
  return xlsxBytesToS3Csvs(buffer, connectionName, companyId, mode, schemaName, createdKeys);
}

/**
 * Parse xlsx bytes, convert each non-empty sheet to CSV, upload to S3.
 * Shared by both CSV uploads (xlsx file already in S3) and Google Sheets import.
 * createdKeys is a caller-owned array; each successfully stored parquet key is pushed
 * onto it so the caller can clean up on failure.
 */
async function xlsxBytesToS3Csvs(
  buffer: Buffer,
  connectionName: string,
  companyId: number,
  mode: string,
  schemaName: string,
  createdKeys: string[] = [],
): Promise<IncomingFile[]> {
  const store = createObjectStore();
  const workbook = XLSX.read(buffer, { type: 'buffer' });
  const results: IncomingFile[] = [];

  // One DuckDB instance per call — no S3 config needed (local file ops only)
  const instance = await DuckDBInstance.create(':memory:');
  const conn = await instance.connect();
  try {
    for (const sheetName of workbook.SheetNames) {
      const ws = workbook.Sheets[sheetName];
      const csvData = XLSX.utils.sheet_to_csv(ws, { blankrows: false });
      if (!csvData.trim()) continue; // skip empty sheets

      const safeName =
        sheetName.toLowerCase().replace(/[^a-z0-9]+/g, '_').replace(/^_+|_+$/g, '') || 'sheet';
      const csvFilename = `${safeName}.csv`;
      const uuid = randomUUID();
      const tmpCsvPath     = join(tmpdir(), `${uuid}.csv`);
      const tmpParquetPath = join(tmpdir(), `${uuid}.parquet`);
      const storageKey = `${companyId}/csvs/${mode}/${connectionName}/${uuid}.parquet`;

      try {
        writeFileSync(tmpCsvPath, csvData, 'utf-8');
        await conn.run(
          `COPY (SELECT * FROM read_csv_auto('${tmpCsvPath}')) TO '${tmpParquetPath}' (FORMAT PARQUET, COMPRESSION ZSTD)`
        );
        const parquetBuffer = readFileSync(tmpParquetPath);
        await store.put(storageKey, parquetBuffer, 'application/octet-stream');
        // Track only after the put succeeds — caller uses this list for cleanup on failure
        createdKeys.push(storageKey);
        results.push({ filename: csvFilename, s3_key: storageKey, schema_name: schemaName, file_format: 'parquet' });
      } finally {
        try { unlinkSync(tmpCsvPath); } catch { /* ignore */ }
        try { unlinkSync(tmpParquetPath); } catch { /* ignore */ }
      }
    }
  } finally {
    conn.closeSync();
    instance.closeSync();
  }

  if (results.length === 0) throw new Error('No non-empty sheets found in xlsx file');
  return results;
}

// ─── S3 delete ────────────────────────────────────────────────────────────────

/** Delete all stored files under a connection's prefix. Returns true if any were deleted. */
export async function deleteConnectionFiles(
  companyId: number,
  mode: string,
  connectionName: string,
): Promise<boolean> {
  if (isLocalObjectStore()) {
    const dir = resolve(join(LOCAL_UPLOAD_PATH, `${companyId}/csvs/${mode}/${connectionName}`));
    try {
      rmSync(dir, { recursive: true, force: true });
      return true;
    } catch {
      return false;
    }
  }

  if (!OBJECT_STORE_BUCKET) return false;
  const s3 = makeS3();
  const prefix = `${companyId}/csvs/${mode}/${connectionName}/`;

  const keys: string[] = [];
  let token: string | undefined;
  do {
    const res = await s3.send(
      new ListObjectsV2Command({ Bucket: OBJECT_STORE_BUCKET, Prefix: prefix, ContinuationToken: token }),
    );
    for (const obj of res.Contents ?? []) if (obj.Key) keys.push(obj.Key);
    token = res.NextContinuationToken;
  } while (token);

  if (keys.length === 0) return false;

  for (let i = 0; i < keys.length; i += 1000) {
    await s3.send(
      new DeleteObjectsCommand({
        Bucket: OBJECT_STORE_BUCKET,
        Delete: { Objects: keys.slice(i, i + 1000).map(Key => ({ Key })) },
      }),
    );
  }
  return true;
}

// ─── S3 single-file delete ────────────────────────────────────────────────────

/** Delete a single stored file by key. */
export async function deleteS3File(s3Key: string): Promise<void> {
  if (isLocalObjectStore()) {
    const filePath = resolve(join(LOCAL_UPLOAD_PATH, s3Key));
    try { unlinkSync(filePath); } catch (err: any) {
      if (err.code !== 'ENOENT') throw err;
    }
    return;
  }
  if (!OBJECT_STORE_BUCKET) throw new Error('OBJECT_STORE_BUCKET is not configured');
  const s3 = makeS3();
  await s3.send(new DeleteObjectCommand({ Bucket: OBJECT_STORE_BUCKET, Key: s3Key }));
}

// ─── DuckDB metadata extraction ───────────────────────────────────────────────

async function configureDuckDBForS3(conn: Awaited<ReturnType<InstanceType<typeof DuckDBInstance>['connect']>>) {
  await conn.run('INSTALL httpfs');
  await conn.run('LOAD httpfs');
  await conn.run(`SET s3_region = '${OBJECT_STORE_REGION}'`);
  if (OBJECT_STORE_ACCESS_KEY_ID) await conn.run(`SET s3_access_key_id = '${OBJECT_STORE_ACCESS_KEY_ID}'`);
  if (OBJECT_STORE_SECRET_ACCESS_KEY) await conn.run(`SET s3_secret_access_key = '${OBJECT_STORE_SECRET_ACCESS_KEY}'`);
  if (OBJECT_STORE_ENDPOINT) {
    await conn.run(`SET s3_endpoint = '${OBJECT_STORE_ENDPOINT}'`);
    await conn.run("SET s3_url_style = 'path'");
  }
}

async function readFileMetadata(
  conn: Awaited<ReturnType<InstanceType<typeof DuckDBInstance>['connect']>>,
  s3Url: string,
  format: 'csv' | 'parquet',
): Promise<{ rowCount: number; columns: Array<{ name: string; type: string }> }> {
  const viewName = `__meta_${randomUUID().replace(/-/g, '_')}`;
  const readExpr = format === 'parquet' ? `read_parquet('${s3Url}')` : `read_csv_auto('${s3Url}')`;

  await conn.run(`CREATE OR REPLACE TEMP VIEW "${viewName}" AS SELECT * FROM ${readExpr}`);

  const countResult = await conn.run(`SELECT COUNT(*) AS cnt FROM "${viewName}"`);
  const countRows = await countResult.getRowObjectsJS() as Array<{ cnt: bigint | number }>;
  const rowCount = Number(countRows[0]?.cnt ?? 0);

  const descResult = await conn.run(`DESCRIBE "${viewName}"`);
  const descRows = await descResult.getRowObjectsJS() as Array<{ column_name: string; column_type: string }>;
  const columns = descRows.map(r => ({ name: r.column_name, type: r.column_type }));

  await conn.run(`DROP VIEW IF EXISTS "${viewName}"`);
  return { rowCount, columns };
}

// ─── CSV → Parquet conversion ─────────────────────────────────────────────────

type DuckConn = Awaited<ReturnType<InstanceType<typeof DuckDBInstance>['connect']>>;

/**
 * Convert a CSV to Parquet (same key prefix, `.parquet` extension).
 * Works with both S3 and local filesystem storage.
 * Deletes the original CSV on success.
 * Returns the new key, or null if conversion fails (original left intact).
 */
async function convertCsvToParquet(conn: DuckConn, csvKey: string): Promise<string | null> {
  const parquetKey = csvKey.replace(/\.[^.]+$/, '') + '.parquet';
  const csvUrl     = getStorageUrl(csvKey);
  const parquetUrl = getStorageUrl(parquetKey);
  try {
    if (isLocalObjectStore()) {
      mkdirSync(dirname(parquetUrl), { recursive: true });
    }
    await conn.run(
      `COPY (SELECT * FROM read_csv_auto('${csvUrl}')) TO '${parquetUrl}' (FORMAT PARQUET, COMPRESSION ZSTD)`
    );
    // Parquet written — remove the now-redundant CSV
    if (isLocalObjectStore()) {
      try { unlinkSync(csvUrl); } catch { /* ignore */ }
    } else {
      await makeS3().send(new DeleteObjectCommand({ Bucket: OBJECT_STORE_BUCKET!, Key: csvKey }));
    }
    return parquetKey;
  } catch (err) {
    console.warn(`[csv-processor] Parquet conversion failed for ${csvKey}:`, err);
    // Delete any partial parquet DuckDB may have written before failing
    if (isLocalObjectStore()) {
      try { unlinkSync(parquetUrl); } catch { /* ignore — may not exist */ }
    } else {
      try {
        await makeS3().send(new DeleteObjectCommand({ Bucket: OBJECT_STORE_BUCKET!, Key: parquetKey }));
      } catch { /* ignore */ }
    }
    return null; // keep CSV as fallback
  }
}

// ─── Main registration function ───────────────────────────────────────────────

/**
 * Process a list of files (already uploaded to S3):
 * - Expands any xlsx files into per-sheet CSVs
 * - Reads row count + column metadata for each file via DuckDB
 * - Returns enriched file records ready to store in the connection config
 *
 * Atomic guarantee: either ALL files are fully written and DuckDB-validated and
 * the full results list is returned, OR every file created during this call is
 * deleted before throwing. No orphaned files are left on disk/S3.
 */
export async function processFilesFromS3(
  companyId: number,
  mode: string,
  connectionName: string,
  incomingFiles: IncomingFile[],
): Promise<RegisteredFile[]> {
  if (!isLocalObjectStore() && !OBJECT_STORE_BUCKET) throw new Error('OBJECT_STORE_BUCKET is not configured');

  // Every key created in this call is tracked here for atomic cleanup on failure.
  const toCleanup: string[] = [];

  try {
    // Expand xlsx files into CSV sheets
    const flatFiles: IncomingFile[] = [];
    for (const file of incomingFiles) {
      const fmt = file.file_format ?? detectFileFormat(file.filename);
      if (fmt === 'xlsx') {
        toCleanup.push(file.s3_key); // original xlsx uploaded by client
        const sheets = await expandXlsxFromS3(
          file.s3_key, connectionName, companyId, mode, file.schema_name ?? 'public',
          toCleanup, // xlsxBytesToS3Csvs pushes each sheet parquet key here after a successful put
        );
        flatFiles.push(...sheets);
      } else {
        toCleanup.push(file.s3_key); // original CSV/parquet uploaded by client
        flatFiles.push({ ...file, file_format: fmt });
      }
    }

    // Auto-assign table names where not provided, ensuring uniqueness
    const noNameFiles = flatFiles.filter(f => !f.table_name).map(f => f.filename);
    const autoNames = ensureUniqueTableNames(noNameFiles);

    const explicitNames = new Set(flatFiles.filter(f => f.table_name).map(f => f.table_name!));
    for (const autoName of autoNames.values()) {
      if (explicitNames.has(autoName)) {
        throw new Error(`Table name collision: '${autoName}' conflicts with a user-supplied table name`);
      }
    }

    // Convert CSV → Parquet and extract metadata via DuckDB
    const instance = await DuckDBInstance.create(':memory:');
    const conn = await instance.connect();
    try {
      if (!isLocalObjectStore()) {
        await configureDuckDBForS3(conn);
      }

      const results: RegisteredFile[] = [];
      for (const file of flatFiles) {
        let s3Key = file.s3_key;
        let format = (file.file_format ?? 'csv') as 'csv' | 'parquet';
        const tableName = file.table_name ?? autoNames.get(file.filename) ?? sanitizeTableName(file.filename);

        // Convert CSV → Parquet for fast columnar queries; fall back to CSV on error
        if (format === 'csv') {
          const converted = await convertCsvToParquet(conn, s3Key);
          if (converted) {
            // CSV was deleted by convertCsvToParquet; swap tracking to the parquet key
            const idx = toCleanup.lastIndexOf(s3Key);
            if (idx >= 0) toCleanup[idx] = converted;
            else toCleanup.push(converted);
            s3Key = converted;
            format = 'parquet';
          }
          // conversion returned null → original CSV key stays in toCleanup
        }

        const fileUrl = getStorageUrl(s3Key);
        const { rowCount, columns } = await readFileMetadata(conn, fileUrl, format);
        results.push({
          table_name: tableName,
          schema_name: file.schema_name ?? 'public',
          s3_key: s3Key,
          file_format: format,
          filename: file.filename,
          row_count: rowCount,
          columns,
        });
      }
      return results;
    } finally {
      conn.closeSync();
      instance.closeSync();
    }
  } catch (err) {
    // Atomic rollback: remove every file created during this call before surfacing the error
    await Promise.allSettled(toCleanup.map(key => deleteS3File(key)));
    throw err;
  }
}

// ─── Google Sheets helpers ────────────────────────────────────────────────────

export function parseSpreadsheetId(url: string): string {
  const match = url.match(/\/spreadsheets\/d\/([a-zA-Z0-9-_]+)/);
  if (!match) throw new Error(`Cannot parse spreadsheet ID from URL: ${url}`);
  return match[1];
}

export async function downloadSpreadsheetAsXlsx(spreadsheetId: string): Promise<Buffer> {
  const exportUrl = `https://docs.google.com/spreadsheets/d/${spreadsheetId}/export?format=xlsx`;
  const res = await fetch(exportUrl, { signal: AbortSignal.timeout(30_000) });
  if (res.status === 404) throw new Error('Spreadsheet not found — it may be private or deleted');
  if (res.status === 401 || res.status === 403) throw new Error('Spreadsheet is not publicly accessible');
  if (!res.ok) throw new Error(`Failed to download spreadsheet: ${res.status} ${res.statusText}`);
  return Buffer.from(await res.arrayBuffer());
}

/** Download a Google Sheet, expand sheets to CSVs, upload all to S3. */
export async function importGoogleSheetToS3(
  spreadsheetUrl: string,
  connectionName: string,
  companyId: number,
  mode: string,
  schemaName: string,
): Promise<{ files: IncomingFile[]; spreadsheetId: string }> {
  const spreadsheetId = parseSpreadsheetId(spreadsheetUrl);
  const xlsxBuffer = await downloadSpreadsheetAsXlsx(spreadsheetId);
  const files = await xlsxBytesToS3Csvs(xlsxBuffer, connectionName, companyId, mode, schemaName);
  return { files, spreadsheetId };
}

// ─── File format detection ─────────────────────────────────────────────────────

export function detectFileFormat(filename: string): 'csv' | 'parquet' | 'xlsx' {
  const ext = filename.split('.').pop()?.toLowerCase();
  if (ext === 'parquet' || ext === 'pq') return 'parquet';
  if (ext === 'xlsx') return 'xlsx';
  return 'csv';
}
